src/coreclr/src/inc/readytorun.h

```diff
@@ -342,6 +344,30 @@ enum ReadyToRunHelper
     READYTORUN_HELPER_StackProbe                = 0x111,
 };
 
+enum ReadyToRunInstructionSet
```

Comment:
Should there be a note about keeping this in sync (generally speaking) with https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/instr.h#L295?

I'd imagine we wouldn't want to add support for a new ISA in one place and forget about adding it in the other.

Reply:
I agree, there is a problem here. There are a thicket of different ways in which various closely related portions of the system treat these instruction sets. I currently count 5 different representations (The string one I introduced throughout crossgen2, the ReadyToRunInstructionSet enum, the CORJIT_Flags, the jit internal instruction set enum, and the managed hardware intrinsics classes.) At the moment, I have logic which mostly enforces that the managed hardware intrinsics and the string names are kept in sync, the logic mapping from jit internal instruction sets to strings is setup to produce noway asserts if there isn't a mapping, and the logic mapping string names to CORJIT flags, produces errors if there is no mapping. Thus there is a decent amount of enforcement, but its really messy, and its likely that any actual attempt to add a new instruction set will be really frustrating as a developer finds more and more enums and such to update.

Reply:
Is it fundamentally necessary that these be distinct? Why can't they be shared in some way?

Reply:
All constants and tables used in the ReadyToRun file format are described in the ReadyToRun file format headers. (And *nothing* else is.) By policy, the common portions of the compiler do not use any ReadyToRun enums or data structures. (Fortunately, the translation between R2R enum and the strings in crossgen is straightforward and transparent.)

More complex is the handling of CORJIT_Flags, and the jit internal instruction set enums. I'd be willing to code up consolidating them on the R2R enum, by removing all of the CORJIT_Flags that are instruction set specific, and either using the R2R enum internally in the JIT, or by making sure the R2R enum values match up with the jit internal enum. (This would involve removing some enum values that have already been defined, under the theory that they cannot be used without feeding them through the rest of the system in any case.) I'd be happy to do this, as the current morass of translations is extremely confusing and risky.

Reply:
How about this...

`compMayUseIsa(isa)`
- Informs the jit that it is allowed to use a particular `isa` in its codegen
- This usage must not change the behavior of the generated application, although it may change its performance characteristics
- It returns true if `isa` is supported, and when operating as an AOT compiler records into the R2R image that the `isa` must be supported at runtime of the generated code.
- It returns false if `isa` is not supported. This false is not recorded with the AOT compiler as the lack of use of the instruction set is not to change the behavior of the application.

Example usages:
When choosing to generate the code for a call to the Math.Round function, the JIT may choose to generate a particular optimized sequence of assembly instructions if SSE4.1 is available, but will otherwise call the Math.Round function. In this case, the compMayUseIsa function is appropriate to use.
Example usages which are not permitted:
When generating the code for Sse42.IsSupported, the compMayUseIsa function must not be used. If the application has code that uses Sse42.IsSupported, the return value of the IsSupported intrinsic is semantically meaningful, and changes in its value are visible in the state of the program. Thus compMayUseIsa must not be used.

`compUsesOrDoesNotUseIsa(isa)`
- Informs the jit that it is allowed to use a particular
- This usage may change the behavior of the generated application, as in changing the meaning of Vector<T>.Size, or the return value of Sse42.IsSupported.
- It returns true if `isa` is supported, and when operating as an AOT compiler records into the R2R image that the `isa` must be supported at runtime of the generated code.
- It returns false if `isa` is not supported, and when operating as an AOT compiler records into the R2R image that the `isa` must not be supported at runtime of the generated code.

Example usages:
When choosing the behavior of an IsSupported intrinsic, or when generating the code for a platform intrinsic. The behavior of these intrinsics varies based on the presence/absence of support, and thus the runtime must verify that the `isa` is supported exactly as when compiled.

Example usages which are discouraged (although not strictly incorrect):
When choosing to generate the code for a call to the Math.Round function, the JIT may choose to generate a particular optimized sequence of assembly instructions if SSE4.1 is available, but will otherwise call the Math.Round function. compUsesIsa should not be used for this case, as the usage of the instruction set has no bearing on the behavior of the generated code.